dbclient/HiveClient.py

```diff
@@ -426,23 +440,27 @@ def log_all_tables(self, db_name, cid, ec_id, metastore_dir, err_log_path, succe
         batch_size = 100    # batch size to iterate over databases
         num_of_buckets = (num_of_tables // batch_size) + 1     # number of slices of the list to take
 
-        all_tables = []
         with open(success_log_path, 'a') as sfp:
             for m in range(0, num_of_buckets):
                 tables_slice = 'print(all_tables[{0}:{1}])'.format(batch_size*m, batch_size*(m+1))
                 results = self.submit_command(cid, ec_id, tables_slice)
                 table_names = ast.literal_eval(results['data'])
                 for table_name in table_names:
                     print("Table: {0}".format(table_name))
-                    is_successful = self.log_table_ddl(cid, ec_id, db_name, table_name, metastore_dir,
-                                                       err_log_path, has_unicode)
+                    full_table_name = f'{db_name}.{table_name}'
+                    if checkpoint_metastore_set.contains(full_table_name):
+                        is_successful = 0
+                    else:
+                        is_successful = self.log_table_ddl(cid, ec_id, db_name, table_name, metastore_dir,
+                                                           err_log_path, has_unicode)
                     if is_successful == 0:
-                        print(f"Exported {db_name}.{table_name}")
-                        success_item = {'table': f'{db_name}.{table_name}', 'iam': iam}
+                        print(f"Exported {full_table_name}")
+                        success_item = {'table': full_table_name, 'iam': iam}
                         sfp.write(json.dumps(success_item))
                         sfp.write('\n')
                     else:
                         print("Logging failure")
+                    checkpoint_metastore_set.write(full_table_name)
```

Comment:
The same question. Should we checkpoint failed operations (which will prevent retry)? If this is expected, I think we should add a field to indicate success or failure in the checkpoint log.

I could imagine in addition to resume interrupted process, another major use case for checkpointing would be retry on failed operations.

Reply:
This write is dangerous because the write to sfp in line #459 is not flushed yet. If the pipeline restarts after this point, data will be lost.

Reply:
@kevinkim9264 let's take this as an example for the data loss risk I mentioned.

In line #459 and #460, data X is written into sfp but not flushed to underlying file yet because of buffering (https://docs.python.org/3/library/io.html#io.IOBase.flush). In line #463, the record has been logged into checkpoint file and flushed (see discussions in checkpoint_service.py).

If the process restarts right after #463, we will resume the previous progress according to checkpoint file. Then, we will skip logging data X because it exists in checkpoint file. But actually data X is never written into the log file which causes the data loss.

Reply:
Thanks for catching that, the checkpoint write needs to go inside the successful block. Fixed it and added flush to the sfp write. 

Yes, checkpoint service can be used for failure handling, and we can evolve the checkpoint service to handle that at a later point of time. The first iteration can be used just for resuming interruped process.

Reply:
ah yeah, basically we need to make sure the write is "fully" done before we ever checkpoint. Once we ensure that, though, we should be safe from the data loss risk?

Reply:
That's correct, all data to be flushed to disk before checkpointing happens

Reply:
When we provide a feature or a library, in general, we should make it easy to use correctly and hard to use in a wrong way. However, the CheckpointService is opposite by design.

Even though we can use it correctly right now, changes in the future could easily break it, for example, someone thinks the flush is not necessary and remove it. On the other hand, making it perfect will take much longer, e.g. unify all file operations, a reasonable workaround is to over-documenting the potential risks in both library definition and all its references to make sure readers understand the risks.

@souravkhandelwal09 what do you think?

Reply:
What makes CheckpointService easy to use it the wrong way and hard to use it correctly ? Can you elaborate on that ?

Achieving pure fault tolerance will require providing a library of new file operations and overhaul of all the clients using it. That is going to be a massive effort. I kept the design such that its easy to use by the clients without disrupting much of its logic as we don't have test coverage in our clients. Also, I didn't want to spend a ton of time for something which is short lived, developed only by 3-4 committers, and doesn't have more use cases than ST migration. There is always a trade off and I think the risks are acceptable here. I am going to add some comments to highlight the risk of removing flush.

Reply:
Why I said it's easy to use it in the wrong way:
In order to use CheckpointService correctly, you'll need to keep in mind that the order of persistence i.e. flush (not the order of write) are always correct. Imagine if a 4th person who didn't review this PR is going to modify the code, how confident are you that they won't mess it up?

I totally agree with you that documenting is a reasonable workaround as I commented previously.